airgun/views/smart_class_parameter.py

```diff
@@ -0,0 +1,156 @@
+from widgetastic.widget import (
+    Checkbox,
+    ParametrizedLocator,
+    Select,
+    Text,
+    TextInput,
+    View,
+)
+from widgetastic_patternfly import BreadCrumb
+
+from airgun.views.common import BaseLoggedInView, SearchableViewMixin
+from airgun.widgets import SatTable
+
+
+class MatcherAttribute(View):
+    """Represent smart class parameter matcher attribute pair. Usually, it
+    looks like as two fields separated by '=' mark
+    """
+    matcher_attribute_type = Select(
+        ".//select[contains(@class, 'matcher_key')]")
+    matcher_attribute_value = TextInput(
+        locator=".//input[contains(@class, 'matcher_value')]")
+
+
+class SmartClassParameterContent(View):
+    ROOT = ParametrizedLocator('{@locator}')
+    key = TextInput(locator=".//input[contains(@name, '[key]')]")
+    description = TextInput(
+        locator=".//textarea[contains(@name, '[description]')]")
+    puppet_environment = TextInput(
+        locator=".//input[contains(@name, '[environment_classes]')]")
+    puppet_class = TextInput(
+        locator=".//input[contains(@name, '[puppetclass_id]')]")
+    override = Checkbox(
+        locator=".//input[contains(@name, '[override]') and @type!='hidden']")
+    key_type = Select(locator=".//select[contains(@name, '[key_type]')]")
+    default_value = TextInput(
+        locator=".//textarea[contains(@name, '[default_value]')]")
+    omit = Checkbox(
+        locator=".//input[contains(@name, '[omit]') and @type!='hidden']")
+    hidden = Checkbox(
+        locator=".//input[contains(@name, '[hidden_value]') and "
+                "@type!='hidden']"
+    )
+
+    def __init__(self, parent, locator, logger=None):
+        View.__init__(self, parent, logger=logger)
+        self.locator = locator
+
+    @View.nested
+    class optional_input_validators(View):
+        expander = Text(
+            ".//h2[contains(@data-target, '#optional_input_validators_')]")
+        required = Checkbox(
+            locator=".//input[contains(@name, '[required]') and "
+                    "@type!='hidden']"
+        )
+        validator_type = Select(
+            locator=".//select[contains(@name, '[validator_type]')]")
+        validator_rule = TextInput(
+            locator=".//input[contains(@name, '[validator_rule]')]")
+
+        def __init__(self, parent, logger=None):
+            View.__init__(self, parent, logger=logger)
+            if 'collapsed' in self.browser.classes(self.expander):
```

Comment:
yes, because we do not have `before_read`, right? Even if we had, it would be bad idea to duplicate code in different methods, like for `before_read` and `before_fill`. I examined the code and that one approach seems as only one possible